src/Algebra/Properties/CommutativeMonoid.agda

```diff
@@ -0,0 +1,181 @@
+∑-+-hom (suc n) f g =
+  let fz = f zero
+      gz = g zero
+      ∑f  = ∑[ i < n ] f (suc i)
+      ∑g  = ∑[ i < n ] g (suc i)
+      ∑fg = ∑[ i < n ] (f (suc i) + g (suc i))
+  in begin
+    (fz + ∑f) + (gz + ∑g)      ≈⟨ +-assoc _ _ _ ⟩
+    fz + (∑f + (gz + ∑g))      ≈⟨ +-cong refl (sym (+-assoc _ _ _)) ⟩
+    fz + ((∑f + gz) + ∑g)      ≈⟨ +-cong refl (+-cong (+-comm _ _) refl) ⟩
+    fz + ((gz + ∑f) + ∑g)      ≈⟨ +-cong refl (+-assoc _ _ _) ⟩
+    fz + (gz + (∑f + ∑g))      ≈⟨ +-cong refl (+-cong refl (∑-+-hom n _ _)) ⟩
+    fz + (gz + ∑fg)            ≈⟨ sym (+-assoc _ _ _) ⟩
+    fz + gz + ∑fg              ∎
+
+-- The '∑' operator commutes with itself.
+
+∑-comm : ∀ n m (f : Fin n → Fin m → Carrier) → ∑[ i < n ] ∑[ j < m ] f i j ≈ ∑[ j < m ] ∑[ i < n ] f i j
+∑-comm zero m f = sym (sumₜ-zero m)
+∑-comm (suc n) m f =
+  begin
+    ∑[ j < m ] f zero j + ∑[ i < n ] ∑[ j < m ] f (suc i) j   ≈⟨ +-cong refl (∑-comm n m _) ⟩
+    ∑[ j < m ] f zero j + ∑[ j < m ] ∑[ i < n ] f (suc i) j   ≈⟨ ∑-+-hom m _ _ ⟩
+    ∑[ j < m ] (f zero j + ∑[ i < n ] f (suc i) j)            ∎
+
+-- Any permutation of a table has the same sum as the original.
+
+sumₜ-permute : ∀ {n} t (π : Permutation′ n) → sumₜ t ≈ sumₜ (rearrange (π ⟨$⟩ʳ_) t)
+sumₜ-permute {zero} t π = refl
+sumₜ-permute {suc n} t π =
+  let f = lookup t
+      0i = zero
+      ππ0 = π ⟨$⟩ʳ (π ⟨$⟩ˡ 0i)
+  in begin
+    sumₜ t                                                                      ≡⟨⟩
+    f 0i + sumₜ (rearrange (punchIn 0i) t)                                      ≈⟨ +-cong refl (sumₜ-permute _ (Perm.remove (π ⟨$⟩ˡ 0i) π)) ⟩
+    f 0i + sumₜ (rearrange (punchIn 0i ∘ (Perm.remove (π ⟨$⟩ˡ 0i) π ⟨$⟩ʳ_)) t)  ≡⟨ P.cong₂ _+_ P.refl (sumₜ-cong≡ (P.cong f ∘ P.sym ∘ Perm.punchIn-permute′ π 0i)) ⟩
+    f 0i + sumₜ (rearrange ((π ⟨$⟩ʳ_) ∘ punchIn (π ⟨$⟩ˡ 0i)) t)                 ≡⟨ P.cong₂ _+_ (P.cong f (P.sym (Perm.inverseʳ π))) P.refl ⟩
+    f _  + sumₜ (rearrange ((π ⟨$⟩ʳ_) ∘ punchIn (π ⟨$⟩ˡ 0i)) t)                 ≈⟨ sym (sumₜ-punchIn (rearrange (π ⟨$⟩ʳ_) t) (π ⟨$⟩ˡ 0i)) ⟩
+    sumₜ (rearrange (π ⟨$⟩ʳ_) t)                                                ∎
+
+-- A version of 'sumₜ-permute' allowing heterogeneous sum lengths.
+
+sumₜ-permute′ : ∀ {m n} t (π : Permutation m n) → sumₜ t ≈ sumₜ (rearrange (π ⟨$⟩ʳ_) t)
+sumₜ-permute′ t π with Perm.↔⇒≡ π
+sumₜ-permute′ t π | P.refl = sumₜ-permute t π
+
+∑-permute : ∀ {n} f (π : Permutation′ n) → ∑[ i < n ] f i ≈ ∑[ i < n ] f (π ⟨$⟩ʳ i)
+∑-permute = sumₜ-permute ∘ tabulate
+
+∑-permute′ : ∀ {m n} f (π : Permutation m n) → ∑[ i < n ] f i ≈ ∑[ i < m ] f (π ⟨$⟩ʳ i)
+∑-permute′ = sumₜ-permute′ ∘ tabulate
+
+-- If the function takes the same value at 'i' and 'j', then transposing 'i' and
+-- 'j' then selecting 'j' is the same as selecting 'i'.
+
+select-transpose : ∀ {n} t (i j : Fin n) → lookup t i ≈ lookup t j → ∀ k → (lookup (select 0# j t) ∘ PermC.transpose i j) k ≈ lookup (select 0# i t) k
+select-transpose _ i j e k with k FP.≟ i
```

Comment:
You could use `...` notation with the `with` here to remove quite a lot of text in this proof.